contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol

```diff
@@ -0,0 +1,446 @@
+    uint256 paymentNoFee = (1e18 * gasPrice * executionGas) / uint256(weiPerUnitLink);
+    uint256 fee = uint256(donFee) + uint256(adminFee);
+    if (paymentNoFee > (1e27 - fee)) {
+      revert PaymentTooLarge(); // Payment + fee cannot be more than all of the link in existence.
+    }
+    return uint96(paymentNoFee + fee);
+  }
+
+  // ================================================================
+  // |                       Billing Methods                        |
+  // ================================================================
+  /**
+   * @notice Initiate the billing process for an Functions request
+   * @param data Encoded Chainlink Functions request data, use FunctionsClient API to encode a request
+   * @param billing Billing configuration for the request
+   * @return requestId - A unique identifier of the request. Can be used to match a request to a response in fulfillRequest.
+   * @return estimatedCost
+   * @return gasAfterPaymentCalculation
+   * @return requestTimeoutSeconds
+   * @dev Only callable by the Functions Router
+   */
+  function _startBilling(
+    bytes memory data,
+    RequestBilling memory billing
+  ) internal returns (bytes32, uint96, uint256, uint256) {
+    // No lower bound on the requested gas limit. A user could request 0
+    // and they would simply be billed for the gas and computation.
+    if (billing.callbackGasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(billing.callbackGasLimit, s_config.maxGasLimit);
+    }
+
+    // Check that subscription can afford the estimated cost
+    uint96 donFee = getDONFee(data, billing);
+    uint96 adminFee = getAdminFee(data, billing);
+    uint96 estimatedCost = _calculateCostEstimate(billing.callbackGasLimit, billing.expectedGasPrice, donFee, adminFee);
+    IFunctionsSubscriptions subscriptions = IFunctionsSubscriptions(address(s_router));
+    (uint96 balance, uint96 blockedBalance, , , ) = subscriptions.getSubscription(billing.subscriptionId);
+    (, uint64 initiatedRequests, ) = subscriptions.getConsumer(billing.client, billing.subscriptionId);
+
+    if (balance - blockedBalance < estimatedCost) {
+      revert InsufficientBalance();
+    }
+
+    bytes32 requestId = computeRequestId(address(this), billing.client, billing.subscriptionId, initiatedRequests + 1);
+
+    Commitment memory commitment = Commitment(
+      billing.subscriptionId,
+      billing.client,
+      billing.callbackGasLimit,
+      billing.expectedGasPrice,
+      address(this),
+      donFee,
+      adminFee,
+      estimatedCost,
+      s_config.gasOverhead + s_config.gasAfterPaymentCalculation,
+      block.timestamp
+    );
+    s_requestCommitments[requestId] = commitment;
+
+    emit BillingStart(requestId, commitment);
+
+    return (requestId, estimatedCost, s_config.gasAfterPaymentCalculation, s_config.requestTimeoutSeconds);
+  }
+
+  /**
+   * @notice Generate a keccak hash request ID
+   */
+  function computeRequestId(
```

Comment:
Does this function still have the problem of potentially creating duplicate request IDs if a consumer deletes & re-adds the same consumer contract? Do we need to add a global nonce?

Reply:
Nope, because the nonce (request # for that consumer) is never reset